File: android/nativepreview/src/main/java/com/travelplaner/nativepreview/data/TripRepository.kt
package com.travelplaner.nativepreview.data

import androidx.room.withTransaction
import com.travelplaner.nativepreview.domain.*
import kotlinx.serialization.json.*
import java.util.UUID
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

interface TripRepository {
    suspend fun list(): List<TripDocument>
    suspend fun create(draft: TripDraft, id: String): TripDocument
    suspend fun edit(id: String, draft: TripDraft): TripDocument
    suspend fun applyMemoryChange(tripID: String, change: MemoryChange): TripDocument
    suspend fun applySchedule(tripID: String, change: ScheduleChange): TripDocument
    suspend fun upsertExpense(tripID: String, expenseID: String, fields: JsonObject): TripDocument
    suspend fun deleteExpense(tripID: String, expenseID: String): TripDocument
    suspend fun updateBudgetSettings(tripID: String, values: JsonObject): TripDocument
    suspend fun updateSettlementParticipants(tripID: String, participants: JsonArray): TripDocument
    suspend fun prepareImport(candidate: ImportCandidate): ImportPreview
    suspend fun commitImport(preview: ImportPreview, decision: ImportDecision): ImportOutcome
    suspend fun listSavedPlaces(): List<SavedPlaceDocument>
    suspend fun savePlace(selection: PlaceSelection, draft: PlaceDraft, id: String): SavedPlaceDocument
    suspend fun deleteSavedPlace(id: String)
}

class RoomTripRepository(private val database: TripDatabase) : TripRepository {
    override suspend fun listSavedPlaces(): List<SavedPlaceDocument> = withContext(Dispatchers.IO) {
        database.savedPlaces().list().map(::savedDocument)
    }
    override suspend fun savePlace(selection: PlaceSelection, draft: PlaceDraft, id: String): SavedPlaceDocument = withContext(Dispatchers.IO) {
        database.withTransaction {
            val payload = selection.durablePayload(id,draft); val source = selection.sourceKey(id)
            val dao = database.savedPlaces()
            dao.find(id)?.let {
                require(it.sourceKey == source) { "같은 저장 요청의 내용이 변경되었습니다." }
                return@withTransaction savedDocument(it)
            }
            dao.findSource(source)?.let { return@withTransaction savedDocument(it) }
            val now = System.currentTimeMillis()
            val row = SavedPlaceEntity(id,source,payload.toString(),now,now)
            dao.insert(row)
            val saved = savedDocument(checkNotNull(dao.find(id)))
            check(saved.payload == payload)
            saved
        }
    }
    override suspend fun deleteSavedPlace(id: String): Unit = withContext(Dispatchers.IO) {
        database.withTransaction { database.savedPlaces().delete(id) }
    }
    private fun savedDocument(row: SavedPlaceEntity): SavedPlaceDocument {
        val payload = Json.parseToJsonElement(row.payload).jsonObject
        check(payload["id"] == JsonPrimitive(row.id)) { "저장된 장소 정보를 읽지 못했어요. 원본은 보존됩니다." }
        return SavedPlaceDocument(row.id,row.sourceKey,payload,row.createdAt,row.updatedAt)
    }
    override suspend fun applySchedule(tripID: String, change: ScheduleChange): TripDocument = withContext(Dispatchers.IO) {
        database.withTransaction {
            val trip = database.trips().find(tripID)?.let(::document) ?: throw IllegalArgumentException("여행을 찾을 수 없어요.")
            store(TripSchedule.apply(change, trip))
        }
    }
    override suspend fun applyMemoryChange(tripID: String, change: MemoryChange): TripDocument = withContext(Dispatchers.IO) {
        database.withTransaction {
            val current = database.trips().find(tripID)?.let(::document) ?: throw IllegalArgumentException("여행을 찾을 수 없어요.")
            val next = TripMemory.apply(change, current, System.currentTimeMillis())
            if (next.json == current.json) current else store(next)
        }
    }
    override suspend fun upsertExpense(tripID: String, expenseID: String, fields: JsonObject): TripDocument = withContext(Dispatchers.IO) {
        database.withTransaction {
            val current = database.trips().find(tripID)?.let(::document) ?: throw IllegalArgumentException("여행을 찾을 수 없어요.")
            val entries = current.json["expenses"]?.jsonArray.orEmpty()
            val merged = entries.map { element ->
                val objectValue = element.jsonObject
                if (objectValue["id"]?.jsonPrimitive?.content == expenseID) JsonObject(objectValue + fields + ("id" to JsonPrimitive(expenseID))) else objectValue
            }.toMutableList()
            if (merged.none { it.jsonObject["id"]?.jsonPrimitive?.content == expenseID }) merged += JsonObject(fields + ("id" to JsonPrimitive(expenseID)))
            return@withTransaction store(updatedRaw(current, "expenses" to JsonArray(merged)))
        }
    }
    override suspend fun deleteExpense(tripID: String, expenseID: String): TripDocument = withContext(Dispatchers.IO) {
        database.withTransaction {
            val current = database.trips().find(tripID)?.let(::document) ?: throw IllegalArgumentException("여행을 찾을 수 없어요.")
            val entries = current.json["expenses"]?.jsonArray.orEmpty()
            store(updatedRaw(current, "expenses" to JsonArray(entries.filterNot { it.jsonObject["id"]?.jsonPrimitive?.content == expenseID })))
        }
    }
    override suspend fun updateBudgetSettings(tripID: String, values: JsonObject): TripDocument = withContext(Dispatchers.IO) {
        database.withTransaction {
            val current = database.trips().find(tripID)?.let(::document) ?: throw IllegalArgumentException("여행을 찾을 수 없어요.")
            val settings = current.json["budgetSettings"]?.jsonObject ?: buildJsonObject { }
            store(updatedRaw(current, "budgetSettings" to JsonObject(settings + values)))
        }
    }
    override suspend fun updateSettlementParticipants(tripID: String, participants: JsonArray): TripDocument = withContext(Dispatchers.IO) {
        database.withTransaction {
            val current = database.trips().find(tripID)?.let(::document) ?: throw IllegalArgumentException("여행을 찾을 수 없어요.")
            store(updatedRaw(current, "settlementParticipants" to participants))
        }
    }
    override suspend fun prepareImport(candidate: ImportCandidate): ImportPreview = withContext(Dispatchers.IO) {
        database.withTransaction { preview(candidate) }
    }
    private suspend fun preview(candidate: ImportCandidate): ImportPreview {
        val receipt = database.receipts().find(candidate.sourceHash)
        val lookup = receipt?.targetID ?: candidate.trip.id
        val existing = database.trips().find(lookup)
        val disposition = when {
            receipt != null -> if (existing != null) ImportDisposition.AlreadyImported else ImportDisposition.PreviouslyDeleted
            existing == null -> ImportDisposition.NewTrip
            document(existing).json == candidate.trip.json -> ImportDisposition.IdenticalExisting
            else -> ImportDisposition.Conflict
        }
        val target = if (disposition in listOf(ImportDisposition.Conflict, ImportDisposition.PreviouslyDeleted)) UUID.randomUUID().toString() else lookup
        return ImportPreview(candidate, disposition, target, existing?.json?.encodeToByteArray()?.let(TripBackup::hash), receipt?.let(::receiptHash))
    }
    override suspend fun commitImport(preview: ImportPreview, decision: ImportDecision): ImportOutcome = withContext(Dispatchers.IO) {
        database.withTransaction {
            val fresh = preview(preview.candidate)
            require(fresh.disposition == preview.disposition && fresh.expectedExistingHash == preview.expectedExistingHash && fresh.expectedReceiptHash == preview.expectedReceiptHash) { "내용이 바뀌어 다시 확인이 필요합니다. 파일을 다시 선택해주세요." }
            val disposition = preview.disposition
            if (disposition == ImportDisposition.AlreadyImported) {
                check(fresh.targetID == preview.targetID)
                return@withTransaction ImportOutcome(fresh.targetID, created = false, alreadyImported = true)
            }
            require((disposition in listOf(ImportDisposition.NewTrip, ImportDisposition.IdenticalExisting) && decision == ImportDecision.ConfirmNew) ||
                (disposition == ImportDisposition.Conflict && decision == ImportDecision.KeepBoth) ||
                (disposition == ImportDisposition.PreviouslyDeleted && decision == ImportDecision.RestoreDeleted)) { "기존 여행을 보존하기 위해 가져오기 방법을 확인해주세요." }
            val trip = TripDocument.parse(JsonObject(preview.candidate.trip.json + ("id" to JsonPrimitive(preview.targetID))).toString())
            if (disposition != ImportDisposition.IdenticalExisting) {
                require(database.trips().find(trip.id) == null) { "여행이 변경되었어요. 다시 확인해주세요." }
                store(trip)
            }
            val receipts = database.receipts(); val old = receipts.find(preview.candidate.sourceHash)
            val previous = old?.previousTargetIDs?.let { Json.parseToJsonElement(it).jsonArray.toMutableList() } ?: mutableListOf()
            if (disposition == ImportDisposition.PreviouslyDeleted && old != null) previous.add(JsonPrimitive(old.targetID))
            val receipt = ImportReceiptEntity(preview.candidate.sourceHash, preview.candidate.sourceBytes, trip.id, JsonArray(previous).toString(), System.currentTimeMillis())
            if (old == null) receipts.insert(receipt) else check(receipts.update(receipt) == 1)
            check(document(checkNotNull(database.trips().find(trip.id))).json == trip.json) { "저장한 여행을 확인하지 못했어요." }
            check(receipts.find(receipt.sourceHash)?.sourceBytes?.contentEquals(receipt.sourceBytes) == true) { "백업 원본을 확인하지 못했어요." }
            ImportOutcome(trip.id, disposition != ImportDisposition.IdenticalExisting, alreadyImported = false)
        }
    }
    private fun receiptHash(receipt: ImportReceiptEntity): String = TripBackup.hash(buildJsonObject {
        put("sourceHash", receipt.sourceHash); put("source", TripBackup.hash(receipt.sourceBytes)); put("targetID", receipt.targetID)
        put("previous", receipt.previousTargetIDs); put("importedAt", receipt.importedAt)
    }.toString().encodeToByteArray())
    override suspend fun list(): List<TripDocument> = withContext(Dispatchers.IO) {
        database.trips().list().map(::document)
    }

    override suspend fun create(draft: TripDraft, id: String): TripDocument = withContext(Dispatchers.IO) {
        database.withTransaction {
            val existing = database.trips().find(id)?.let(::document)
            // A saved operation ID makes retries after commit/process death idempotent.
            store(existing?.edit(draft) ?: TripDocument.create(draft, id))
        }
    }

    override suspend fun edit(id: String, draft: TripDraft): TripDocument = withContext(Dispatchers.IO) {
        database.withTransaction {
            val existing = database.trips().find(id)?.let(::document)
                ?: throw IllegalArgumentException("여행을 찾을 수 없어요. 목록을 다시 불러와 주세요.")
            // Re-read inside the transaction so edits cannot overwrite unseen JSON fields.
            store(existing.edit(draft))
        }
    }

    suspend fun save(trip: TripDocument): TripDocument = withContext(Dispatchers.IO) {
        database.withTransaction { store(trip) }
    }

    private suspend fun store(trip: TripDocument): TripDocument {
        val dao = database.trips()
        val existing = dao.find(trip.id)
        if (existing != null) {
            val original = document(existing)
            require(original.json["createdAt"] == trip.json["createdAt"]) { "기존 여행의 생성 시각을 변경할 수 없어요." }
        }
        val entity = TripEntity(trip.id, trip.toJson(), trip.updatedAt)
        if (existing == null) dao.insert(entity) else check(dao.update(entity) == 1)
        val persisted = document(checkNotNull(dao.find(trip.id)))
        check(persisted.json == trip.json) { "저장한 여행을 다시 확인하지 못했어요." }
        return persisted
    }

    private fun updatedRaw(trip: TripDocument, vararg change: Pair<String, JsonElement>): TripDocument {
        val next = buildJsonObject {
            trip.json.forEach { (key, value) -> put(key, value) }
            change.forEach { (key, value) -> put(key, value) }
            put("updatedAt", JsonPrimitive(maxOf(System.currentTimeMillis(), trip.updatedAt + 1)))
        }
        return TripDocument.parse(next.toString())
    }

    private fun document(entity: TripEntity): TripDocument = TripDocument.parse(entity.json).also {
        check(it.id == entity.id && it.updatedAt == entity.updatedAt) { "여행의 저장 정보가 일치하지 않아요." }
    }
}

File: ios/NativeCore/Sources/NativeCore/TripRepository.swift
import Foundation
import CoreData

public final class TripRepository {
    private let context: NSManagedObjectContext
    private static let accessLock = NSRecursiveLock()
    var beforeReceiptWrite: (() throws -> Void)?

    public init(url: URL) throws {
        Self.accessLock.lock(); defer { Self.accessLock.unlock() }
        try FileManager.default.createDirectory(at: url.deletingLastPathComponent(), withIntermediateDirectories: true)
        try TripStoreModel.migrateIfNeeded(at: url)
        let coordinator = NSPersistentStoreCoordinator(managedObjectModel: TripStoreModel.current)
        try coordinator.addPersistentStore(ofType: NSSQLiteStoreType, configurationName: nil, at: url, options: nil)
        context = NSManagedObjectContext(concurrencyType: .privateQueueConcurrencyType)
        context.persistentStoreCoordinator = coordinator
    }

    public func list() throws -> [TripDocument] {
        try transaction {
            let request = NSFetchRequest<NSManagedObject>(entityName: "TripRecord")
            request.sortDescriptors = [NSSortDescriptor(key: "updatedAt", ascending: false), NSSortDescriptor(key: "id", ascending: true)]
            return try context.fetch(request).map { record in
                guard let data = record.value(forKey: "payload") as? Data else { throw TripError.invalidDocument }
                return try TripDocument(data: data)
            }
        }
    }

    public func save(_ trip: TripDocument) throws {
        try transaction { try write(trip) }
    }

    public func applyMemoryChange(tripID: String, change: MemoryChange, now: Double = floor(Date().timeIntervalSince1970 * 1000)) throws -> TripDocument {
        try transaction {
            guard let row = try find(tripID), let bytes = row.value(forKey: "payload") as? Data else { throw TripError.invalidDocument }
            let current = try TripDocument(data: bytes)
            let next = try TripMemory.apply(change, to: current, now: now)
            if try next.encoded() == current.encoded() { return current }
            try write(next)
            guard let saved = try find(tripID)?.value(forKey: "payload") as? Data else { throw TripError.invalidDocument }
            guard try TripDocument(data: saved).encoded() == next.encoded() else { throw TripError.invalidDocument }
            return try TripDocument(data: saved)
        }
    }

    public func upsertExpense(tripID: String, expenseID: String, fields: [String: Any]) throws -> TripDocument {
        try transaction {
            guard let row = try find(tripID), let bytes = row.value(forKey: "payload") as? Data else { throw TripError.invalidDocument }
            var raw = try TripDocument(data: bytes).raw
            var entries: [[String: Any]]
            if let existing = raw["expenses"] {
                guard let decoded = existing as? [[String: Any]] else { throw TripError.invalidDocument }
                entries = decoded
            } else { entries = [] }
            var merged = fields
            merged["id"] = expenseID
            if let index = entries.firstIndex(where: { ($0["id"] as? String) == expenseID }) {
                var current = entries[index]; merged.forEach { current[$0.key] = $0.value }; entries[index] = current
            } else { entries.append(merged) }
            raw["expenses"] = entries
            raw["updatedAt"] = max(floor(Date().timeIntervalSince1970 * 1000), ((raw["updatedAt"] as? NSNumber)?.doubleValue ?? 0) + 1)
            let next = try TripDocument(data: JSONSerialization.data(withJSONObject: raw))
            try write(next)
            guard let saved = try find(tripID)?.value(forKey: "payload") as? Data else { throw TripError.invalidDocument }
            return try TripDocument(data: saved)
        }
    }

    public func deleteExpense(tripID: String, expenseID: String) throws {
        try transaction {
            guard let row = try find(tripID), let bytes = row.value(forKey: "payload") as? Data else { throw TripError.invalidDocument }
            var raw = try TripDocument(data: bytes).raw
            guard let existing = raw["expenses"] else { return }
            guard let entries = existing as? [[String: Any]] else { throw TripError.invalidDocument }
            raw["expenses"] = entries.filter { ($0["id"] as? String) != expenseID }
            raw["updatedAt"] = max(floor(Date().timeIntervalSince1970 * 1000), ((raw["updatedAt"] as? NSNumber)?.doubleValue ?? 0) + 1)
            try write(TripDocument(data: JSONSerialization.data(withJSONObject: raw)))
        }
    }

    public func updateBudgetSettings(tripID: String, values: [String: Any]) throws -> TripDocument {
        try transaction {
            guard let row = try find(tripID), let bytes = row.value(forKey: "payload") as? Data else { throw TripError.invalidDocument }
            var raw = try TripDocument(data: bytes).raw
            var settings = (raw["budgetSettings"] as? [String: Any]) ?? [:]
            values.forEach { settings[$0.key] = $0.value }
            raw["budgetSettings"] = settings
            raw["updatedAt"] = max(floor(Date().timeIntervalSince1970 * 1000), ((raw["updatedAt"] as? NSNumber)?.doubleValue ?? 0) + 1)
            let next = try TripDocument(data: JSONSerialization.data(withJSONObject: raw)); try write(next)
            guard let saved = try find(tripID)?.value(forKey: "payload") as? Data else { throw TripError.invalidDocument }
            return try TripDocument(data: saved)
        }
    }

    public func updateSettlementParticipants(tripID: String, participants: [[String: Any]]) throws -> TripDocument {
        try transaction {
            guard let row = try find(tripID), let bytes = row.value(forKey: "payload") as? Data else { throw TripError.invalidDocument }
            var raw = try TripDocument(data: bytes).raw
            raw["settlementParticipants"] = participants
            raw["updatedAt"] = max(floor(Date().timeIntervalSince1970 * 1000), ((raw["updatedAt"] as? NSNumber)?.doubleValue ?? 0) + 1)
            let next = try TripDocument(data: JSONSerialization.data(withJSONObject: raw)); try write(next)
            guard let saved = try find(tripID)?.value(forKey: "payload") as? Data else { throw TripError.invalidDocument }
            return try TripDocument(data: saved)
        }
    }

    public func delete(id: String) throws {
        try transaction { if let record = try find(id) { context.delete(record) } }
    }

    public func listSavedPlaces() throws -> [SavedPlaceDocument] {
        try transaction {
            let request = NSFetchRequest<NSManagedObject>(entityName: "SavedPlaceRecord")
            request.sortDescriptors = [NSSortDescriptor(key: "updatedAt", ascending: false), NSSortDescriptor(key: "id", ascending: true)]
            return try context.fetch(request).map(savedDocument)
        }
    }
    public func savePlace(selection: PlaceSelection, draft: PlaceDraft, id: String) throws -> SavedPlaceDocument {
        try transaction {
            let payload = try selection.durablePayload(id: id, draft: draft)
            let sourceKey = try selection.sourceKey(creationID: id)
            if let sameID = try savedRecord(field: "id", value: id) {
                let old = try savedDocument(sameID)
                guard old.sourceKey == sourceKey else { throw PlaceError.conflictingOperation }
                // Same source is already saved: never silently overwrite its user memo.
                return old
            }
            if let duplicate = try savedRecord(field: "sourceKey", value: sourceKey) { return try savedDocument(duplicate) }
            let row = NSEntityDescription.insertNewObject(forEntityName: "SavedPlaceRecord", into: context)
            let now = floor(Date().timeIntervalSince1970 * 1000)
            row.setValue(id, forKey: "id"); row.setValue(sourceKey, forKey: "sourceKey")
            row.setValue(try JSONSerialization.data(withJSONObject: payload, options: [.sortedKeys]), forKey: "payload")
            row.setValue(now, forKey: "createdAt"); row.setValue(now, forKey: "updatedAt")
            return try savedDocument(row)
        }
    }
    public func deleteSavedPlace(id: String) throws {
        try transaction { if let record = try savedRecord(field: "id", value: id) { context.delete(record) } }
    }
    private func savedRecord(field: String, value: String) throws -> NSManagedObject? {
        let request = NSFetchRequest<NSManagedObject>(entityName: "SavedPlaceRecord")
        request.predicate = NSPredicate(format: "%K == %@", field, value); request.fetchLimit = 1
        return try context.fetch(request).first
    }
    private func savedDocument(_ row: NSManagedObject) throws -> SavedPlaceDocument {
        guard let id = row.value(forKey: "id") as? String, let source = row.value(forKey: "sourceKey") as? String,
              let bytes = row.value(forKey: "payload") as? Data, let payload = try JSONSerialization.jsonObject(with: bytes) as? [String: Any],
              payload["id"] as? String == id, let created = row.value(forKey: "createdAt") as? Double,
              let updated = row.value(forKey: "updatedAt") as? Double else { throw TripError.invalidDocument }
        return .init(id: id, sourceKey: source, payload: payload, createdAt: created, updatedAt: updated)
    }

    public func applySchedule(tripID: String, change: ScheduleChange) throws -> TripDocument {
        try transaction {
            guard let bytes = try find(tripID)?.value(forKey: "payload") as? Data else { throw ScheduleError.missingItem }
            let next = try TripSchedule.apply(change, to: TripDocument(data: bytes)); try write(next)
            guard let saved = try find(tripID)?.value(forKey: "payload") as? Data, saved == (try next.encoded()) else { throw TripError.invalidDocument }
            return try TripDocument(data: saved)
        }
    }
    public func prepareImport(_ candidate: ImportCandidate) throws -> ImportPreview {
        try transaction { try preview(candidate) }
    }
    public func commitImport(_ expected: ImportPreview, decision: ImportDecision) throws -> ImportOutcome {
        try transaction {
            let fresh = try preview(expected.candidate)
            guard fresh.disposition == expected.disposition, fresh.expectedExistingHash == expected.expectedExistingHash,
                  fresh.expectedReceiptHash == expected.expectedReceiptHash else { throw BackupError.stale }
            let disposition = expected.disposition
            if disposition == .alreadyImported {
                guard fresh.targetID == expected.targetID else { throw BackupError.stale }
                return .init(tripID: fresh.targetID, created: false, alreadyImported: true)
            }
            guard (disposition == .newTrip || disposition == .identicalExisting) && decision == .confirmNew ||
                    disposition == .conflict && decision == .keepBoth || disposition == .previouslyDeleted && decision == .restoreDeleted else { throw BackupError.decision }
            var raw = expected.candidate.trip.raw
            raw["id"] = expected.targetID
            let trip = try TripDocument(data: JSONSerialization.data(withJSONObject: raw))
            if disposition != .identicalExisting {
                guard try find(trip.id) == nil else { throw BackupError.stale }
                try write(trip)
            }
            try beforeReceiptWrite?()
            let existingReceipt = try receipt(expected.candidate.sourceHash)
            var previous = try history(existingReceipt)
            if disposition == .previouslyDeleted, let target = existingReceipt?.value(forKey: "targetID") as? String { previous.append(target) }
            let row = existingReceipt ?? NSEntityDescription.insertNewObject(forEntityName: "ImportReceipt", into: context)
            row.setValue(expected.candidate.sourceHash, forKey: "sourceHash"); row.setValue(expected.candidate.sourceBytes, forKey: "sourceBytes")
            row.setValue(trip.id, forKey: "targetID"); row.setValue(try JSONSerialization.data(withJSONObject: previous), forKey: "previousTargetIDs")
            row.setValue(floor(Date().timeIntervalSince1970 * 1000), forKey: "importedAt")
            guard let payload = try find(trip.id)?.value(forKey: "payload") as? Data,
                  try TripDocument(data: payload).encoded() == trip.encoded(),
                  try receipt(expected.candidate.sourceHash)?.value(forKey: "sourceBytes") as? Data == expected.candidate.sourceBytes else { throw TripError.invalidDocument }
            return .init(tripID: trip.id, created: disposition != .identicalExisting, alreadyImported: false)
        }
    }
    func receiptSource(hash: String) throws -> Data? { try transaction { try receipt(hash)?.value(forKey: "sourceBytes") as? Data } }
    func previousImportTargets(hash: String) throws -> [String] { try transaction { try history(receipt(hash)) } }
    private func preview(_ candidate: ImportCandidate) throws -> ImportPreview {
        let row = try receipt(candidate.sourceHash)
        let lookup = row?.value(forKey: "targetID") as? String ?? candidate.trip.id
        let bytes = try find(lookup)?.value(forKey: "payload") as? Data
        let existingHash = bytes.map(TripBackup.hash), receiptHash = try row.map(hashReceipt)
        let disposition: ImportDisposition
        if row != nil { disposition = bytes == nil ? .previouslyDeleted : .alreadyImported }
        else if let bytes { disposition = try TripDocument(data: bytes).encoded() == candidate.trip.encoded() ? .identicalExisting : .conflict }
        else { disposition = .newTrip }
        let target = disposition == .conflict || disposition == .previouslyDeleted ? UUID().uuidString : lookup
        return .init(candidate: candidate, disposition: disposition, targetID: target, expectedExistingHash: existingHash, expectedReceiptHash: receiptHash)
    }
    private func write(_ trip: TripDocument) throws {
        let row = try find(trip.id) ?? NSEntityDescription.insertNewObject(forEntityName: "TripRecord", into: context)
        row.setValue(trip.id, forKey: "id"); row.setValue(try trip.encoded(), forKey: "payload"); row.setValue(trip.updatedAt, forKey: "updatedAt")
    }
    private func receipt(_ hash: String) throws -> NSManagedObject? {
        let request = NSFetchRequest<NSManagedObject>(entityName: "ImportReceipt"); request.predicate = NSPredicate(format: "sourceHash == %@", hash); request.fetchLimit = 1
        return try context.fetch(request).first
    }
    private func history(_ receipt: NSManagedObject?) throws -> [String] {
        guard let receipt else { return [] }
        guard let data = receipt.value(forKey: "previousTargetIDs") as? Data, let ids = try JSONSerialization.jsonObject(with: data) as? [String] else { throw TripError.invalidDocument }
        return ids
    }
    private func hashReceipt(_ row: NSManagedObject) throws -> String {
        guard let source = row.value(forKey: "sourceBytes") as? Data else { throw TripError.invalidDocument }
        let fields: [String: Any] = ["sourceHash": row.value(forKey: "sourceHash")!, "source": TripBackup.hash(source), "targetID": row.value(forKey: "targetID")!, "previous": try history(row), "importedAt": row.value(forKey: "importedAt")!]
        return TripBackup.hash(try JSONSerialization.data(withJSONObject: fields, options: [.sortedKeys]))
    }
    private func transaction<T>(_ body: () throws -> T) throws -> T {
        Self.accessLock.lock(); defer { Self.accessLock.unlock() }
        return try context.performAndWait {
            context.reset()
            do { let result = try body(); if context.hasChanges { try context.save() }; return result }
            catch { context.rollback(); throw error }
        }
    }

    private func find(_ id: String) throws -> NSManagedObject? {
        let request = NSFetchRequest<NSManagedObject>(entityName: "TripRecord")
        request.predicate = NSPredicate(format: "id == %@", id)
        request.fetchLimit = 1
        return try context.fetch(request).first
    }
}
